rust/tools/authorization-logic/src/souffle/souffle_emitter.rs

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2021 The Raksha Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+use crate::{ast::*, souffle::datalog_ir::*};
+use std::collections::HashSet;
+
+pub struct SouffleEmitter {
+    // This is a set of predicates that need to be declared. This set is
+    // populated by side effect while traversing the program body.
+    decls: HashSet<AstPredicate>,
+}
+
+impl<'a> SouffleEmitter {
+    /// The only public method of the souffle emitter is `emit_program`.
```

Comment:
nit (applicable in multiple other cases throughout): the first sentence of doc comment for a function or a struct should generally be written to start with a verb describing it primare action (e.g. "Produces a Souffle code as a `String` when given a DataLog IR (DLIR) program.")

Also, I don't think that stating which functions are public is very useful as it is already indicated in the code itself.

Reply:
Fixed.